File: src/bus.rs
use crate::rom;

const RAM_SIZE: usize = 2048;

#[derive(Debug)]
pub struct CpuBus {
    ram: [u8; RAM_SIZE],
    cartridge: rom::Cartridge,
}

impl CpuBus {
    pub fn new(cartridge: rom::Cartridge) -> Self {
        Self {
            ram: [0; RAM_SIZE],
            cartridge: cartridge,
        }
    }

    pub fn read(&self, address: u16) -> u8 {
        match address {
            0x0000..=0x1FFF => self.ram[(address & 0x07FF) as usize],
            // PPU registers
            0x2000..=0x3FFF => 0,
            // APU and I/O registers
            0x4000..=0x4017 => 0,
            // APU and I/O functionality that is normally disabled
            0x4018..=0x401F => 0,
            // PRG ROM, PRG RAM and mapper registers
            0x4020..=0xFFFF => self.cartridge.read(address),
            _ => panic!("invalid address: {address:#X}"),
        }
    }

    pub fn write(&mut self, address: u16, data: u8) {
        match address {
            0x0000..=0x1FFF => self.ram[address as usize & 0x07FF] = data,
            // PPU registers
            0x2000..=0x3FFF => (),
            // DMA
            0x4014 => (),
            // APU and I/O registers
            0x4000..=0x4017 => (),
            // APU and I/O functionality that is normally disabled
            0x4018..=0x401F => (),
            // PRG ROM, PRG RAM and mapper registers
            0x4020..=0xFFFF => (),
            _ => (),
        }
    }
}

File: src/bus.cpp
#include "bus.h"


CPUMemoryBus::CPUMemoryBus() :
    cartridge()
{
}

void CPUMemoryBus::connect_cartridge(std::shared_ptr<Cartridge> cartridge_)
{
    cartridge = cartridge_;
}

uint8_t CPUMemoryBus::read(uint16_t address) const
{
    if (address >= 0x0000 && address <= 0x1FFF)
    {
        return ram[address & 0x07FF];
    }
    if (address >= 0x2000 && address <= 0x3FFF)
    {
        // PPU register
        return 0;
    }
    if (address >= 0x4000 && address <= 0x4017)
    {
        // APU and I/O registers
        return 0;
    }
    if (address >= 0x4018 && address <= 0x401F)
    {
        // APU and I/O functionality that is normally disabled
        return 0;
    }
    if (address >= 0x4020 && address <= 0xFFFF)
    {
        // # PRG ROM, PRG RAM and mapper registers
        return cartridge->read(address);
    }
    return 0;
}

void CPUMemoryBus::write(uint16_t address, const uint8_t data)
{
    if (address >= 0x0000 && address <= 0x1FFF)
    {
        ram[address & 0x07FF] = data;
        return;
    }
    if (address >= 0x2000 && address <= 0x3FFF)
    {
        // PPU register
        return;
    }
    if (address == 0x4014)
    {
        // DMA
        return;
    }
    if (address >= 0x4000 && address <= 0x4017)
    {
        // APU and I/O registers
        return;
    }
    if (address >= 0x4018 && address <= 0x401F)
    {
        // APU and I/O functionality that is normally disabled
        return;
    }
    if (address >= 0x4020 && address <= 0xFFFF)
    {
        // # PRG ROM, PRG RAM and mapper registers
        cartridge->write(address, data);
    }
}
